### PR TITLE
Default rule

### PR DESCRIPTION
**Pull request by apalm**
I thought I'd put up this PR for discussion as one possible solution to https://github.com/maticzav/graphql-shield/issues/167.

## Summary of Changes

This diff replaces `shield`'s `whitelist` option with a new option, `defaultRule`. If set, `defaultRule` needs to be a `ShieldRule`; if omitted, it defaults to `allow`, which mimics the current behavior of blacklisting by default if `whitelist` is omitted. Since this is a breaking change for people using `whitelist`, a major version bump would be required.

People currently explicitly setting `whitelist` would need to update their code to:

* `{ whitelist: true }`  -> `{ defaultRule: deny }`
* `{ whitelist: false }` -> `{ defaultRule: allow }`

to get the current behavior.

## Motivation

It should be possible to set a default rule other than simple allow/deny. As mentioned in the linked issue, a use case is setting a custom rule on the majority of fields, with some explicit exceptions.

Thoughts?

### CODE REVIEW

**Comment by maticzav** [comment]
@apalm this is fantastic! I think this could solve many edge cases which people are facing right now. I see that you changed `whitelist` to `defaultRule`, could you, however, persist the `whitelist` option and only add the other one. This way we don't have to roll a major release.

I also see that you added tests! 🎉 Don't know why the coverage dropped though, would you mind checking that?

Let me know if you need any help with anything! Once again, huge thanks for this fantastic PR! 🙂

**Comment by apalm** [comment]
@maticzav Thanks for looking!

> persist the whitelist option and only add the other one

Sure, I could do that; however, which would have precedence if both are specified - `whitelist` or `defaultRule`?

For example, if the options specified are:

```ts
{ whitelist: true, defaultRule: allow }
```

would whitelisting be used (meaning `whitelist` has precedence), or would blacklisting be used (meaning `defaultRule` has precedence)?

I guess we could do a union like:

```ts
interface IOptions1 {
  debug?: boolean
  allowExternalErrors?: boolean
  graphiql?: boolean
  fallback?: string | Error
  whitelist?: boolean
}

interface IOptions2 {
  debug?: boolean
  allowExternalErrors?: boolean
  graphiql?: boolean
  fallback?: string | Error
  defaultRule?: ShieldRule
}

export type IOptions = IOptions1 | IOptions2
```

but not sure if you want to go down that route.

> the coverage dropped though

Yup, I'll definitely check up on that.

cc @ellipticaldoor and @dlerman2 because of their involvement in the discussion in #167.

**Comment by maticzav** [comment]
Perfect! 🎉 

I think it's best to keep both in one `IOptions` definition and add a manual check in the `normalizeOptions` function which would throw if both of them are defined. Furthermore, I think we can remove `whitelist` option from the documentation because `defaultRule` far more elegantly and coherently describes its effects.

Maybe we could think whether it would make more sense to name it `fallbackRule`, and also introduce `fallbackError` in addition to `fallback`, and whether we should apply it as `AND` to every rule or just whenever no other rule is specified. Let me know what you think! 🙂

**Comment by apalm** [comment]
@maticzav

> I think it's best to keep both in one IOptions definition and add a manual check in the normalizeOptions function which would throw if both of them are defined. Furthermore, I think we can remove whitelist option from the documentation because defaultRule far more elegantly and coherently describes its effects.

Okay, I will update the PR to do this.

> make more sense to name it fallbackRule

Sure, that's fine with me. Will update the PR.

> introduce fallbackError in addition to fallback

Sorry, it's not clear to me what `fallbackError` would do if specified. Can you please elaborate?

> whether we should apply it as AND to every rule or just whenever no other rule is specified

Is this statement referring to `fallbackRule` or `fallbackError`? If you're referring to `fallbackRule`, no, I don't think we should apply `fallbackRule` as a AND to every rule. Seems like that could get pretty confusing.

**Comment by apalm** [comment]
@maticzav Ah okay, thanks. That makes sense. Are we going to have both `fallback` and `fallbackError` for the time being? Or rename it (breaking change)?

**Comment by maticzav** [comment]
I think it’s best to avoid breaking changes. Let’s just have both. We can remove them when there’s a necessary breakig change coming up.

**Comment by apalm** [comment]
@maticzav Updated. I believe I've addressed all of your comments. Please take another look whenever you get the chance. Thanks!

File: README.md [hunk 1]
```diff
@@ -372,9 +372,9 @@ const permissions = shield({
 })
 ```
 
-### `Global Fallback`
+### `Global Fallback Error`
```
Comment (maticzav) [inline README.md]:
```suggestion
### `Global Error Fallback`
```

File: src/shield.ts [hunk 2]
```diff
@@ -12,22 +13,43 @@ import { generateMiddlewareGeneratorFromRuleTree } from './generator'
...
 function normalizeOptions(options: IOptionsConstructor): IOptions {
+  if (options.whitelist !== undefined && options.fallbackRule !== undefined) {
+    throw new Error(
+      'You specified both `whitelist` and `fallbackRule`. Please use one or the other.',
+    )
+  }
+
+  if (options.fallback !== undefined && options.fallbackError !== undefined) {
+    throw new Error(
+      'You specified both `fallback` and `fallbackError`. Please use one or the other.',
```
Comment (maticzav) [inline src/shield.ts]:
```suggestion
      'You specified both `fallback` and `fallbackError`. Please use only `fallbackError` instead.',
```

**Comment by maticzav** [comment]
@apalm thank you so much for this amazing PR, I see you've added new tests and coverage remains intact, perfect 🎉. I am merging in a second!

**Pull request merged**